kernel/work.c

```diff
@@ -0,0 +1,971 @@
+
+/* Invoked by work thread */
+static void handle_flush(struct k_work *work)
+{
+	struct z_work_flusher *flusher
+		= CONTAINER_OF(work, struct z_work_flusher, work);
+
+	k_sem_give(&flusher->sem);
+}
+
+static inline void init_flusher(struct z_work_flusher *flusher)
+{
+	k_sem_init(&flusher->sem, 0, 1);
+	k_work_init(&flusher->work, handle_flush);
+}
+
+/* List of pending cancellations. */
+static sys_slist_t pending_cancels;
+
+/* Initialize a canceler record and add it to the list of pending
+ * cancels.
+ *
+ * Invoked with work lock held.
+ *
+ * @param canceler the structure used to notify a waiting process.
+ * @param work the work structure that is to be canceled
+ */
+static inline void init_work_cancel(struct z_work_canceller *canceler,
+				    struct k_work *work)
+{
+	k_sem_init(&canceler->sem, 0, 1);
+	canceler->work = work;
+	sys_slist_append(&pending_cancels, &canceler->node);
+}
+
+/* Complete cancellation of a work item and unlock held lock.
+ *
+ * Invoked with work lock held.
+ *
+ * Invoked from a work queue thread.
+ *
+ * Reschedules.
+ *
+ * @param work the work structre that has completed cancellation
+ */
+static void finalize_cancel_locked(struct k_work *work)
+{
+	struct z_work_canceller *wc, *tmp;
+	sys_slist_t cancels;
+
+	/* Clear this first, so released high-priority threads don't
+	 * see it when doing things.
+	 */
+	atomic_clear_bit(&work->flags, K_WORK_CANCELING_BIT);
+
+	/* Move all pending cancels associated with this work item to
+	 * another list, under lock.  (We can't signal the release
+	 * while the lock is held because it's a reschedule point,
+	 * which would release the lock.)
+	 */
+	sys_slist_init(&cancels);
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&pending_cancels, wc, tmp, node) {
+		if (wc->work == work) {
+			sys_slist_remove(&pending_cancels, NULL, &wc->node);
+			k_sem_give(&wc->sem);
+		}
+	}
+}
+
+void k_work_init(struct k_work *work,
+		  k_work_handler_t handler)
+{
+	__ASSERT_NO_MSG(work != NULL);
+	__ASSERT_NO_MSG(handler != 0);
+
+	*work = (struct k_work)Z_WORK_INITIALIZER(handler);
+}
+
+/* Add a flusher work item to the queue.
+ *
+ * Invoked with work lock and queue lock both held.
+ *
+ * Caller must notify queue of pending work.
+ *
+ * @param queue queue on which a work item may appear.
+ * @param work the work item that is either queued or running on @p
+ * queue
+ * @param flusher an uninitialized/unused flusher object
+ */
+static void queue_flusher_locked(struct k_work_q *queue,
+				 struct k_work *work,
+				 struct z_work_flusher *flusher)
+{
+	bool in_list = false;
+	struct k_work *wn;
+
+	/* Determine whether the work item is still queued. */
+	SYS_SLIST_FOR_EACH_CONTAINER(&queue->pending, wn, node) {
+		if (wn == work) {
+			in_list = true;
+			break;
+		}
+	}
+
+	init_flusher(flusher);
+	if (in_list) {
+		sys_slist_insert(&queue->pending, &work->node,
+				 &flusher->work.node);
+	} else {
+		sys_slist_prepend(&queue->pending, &flusher->work.node);
+	}
```

Comment:
why do you have this condition ? I mean what is the difference if the work is in or not in the list ?

Reply:
If the work item we're waiting for is in list, the work item that notifies of completion must *follow* that item.  If it's not in the list, then the work item has already completed, and the notification item can be run immediately.